content/asm_3.md

```diff
@@ -11,243 +14,628 @@ global _start
+
+Once all parameters of both system calls are ready, we can pass them as arguments to print the sum followed by a new line.
+
+Now, let's build our program with the usual commands:
+
+```bash
+$ nasm -f elf64 -o stack.o stack.asm
+$ ld -o stack stack.o
+```
+
+Then, try to run it:
+
+```bash
+$  ./stack
+Error: expected two command line argument
+$ ./stack 5
+Error: expected two command line argument
+$ ./stack 5 10
+15
+```
+
+Works as expected 🎉🎉🎉
+
+
+## Security considerations
+
+As we have seen in this and in the previous posts, the stack is a crucial concept that is used to manage function calls in our programs. Understanding of how the stack memory managed is not only important for writing programs with re-usable functions but also crucial for writing secure programs. The stack often has been a common source of security vulnerabilities, especially in low-level code and assembly routines. When you use `call` and `ret` instructions, the processor doesn’t verify if the return address is valid, but it just simply pops the address and jumps on it. One of the most common problems is the [stack overflow](https://en.wikipedia.org/wiki/Stack_overflow).
+
+Let's take a look at the simple C function (the function is written on C for simplicity):
+
+```C
+#include <stdio.h>
+#include <string.h>
+
+void foo() {
+    char buffer[8];
+
+    printf("Enter text: ");
+
+    gets(buffer);
+}
+
+int main() {
+    foo();
+    printf("Program exited successfully\n");
+    return 0;
+}
+```
+
+If we will try to build this program and run it, we'll see the following error instead of the *Program exited successfully* string:
+
+```bash
+$ ./test
+Enter text: 123456789
+*** stack smashing detected ***: terminated
+Aborted (core dumped)
+```
+
+The reason for this is that we put on the stack the value which is bigger than our 8 bytes buffer. Happily instead of overwriting of return address or segmentation fault error we have got "stack smashing detected" error. This check is done by the modern compiler to prevent overwriting of critical data. There are other techniques in modern compilers and operating system kernels to mitigate vulnerabilities related to stack, like:
```

Comment:
```suggestion
The reason for this error is that we put on the stack a value bigger than our 8-byte buffer. Happily, instead of overwriting the return address or segmentation fault error, we get a `stack smashing detected` error. This check is done by a modern compiler to prevent overwriting critical data. There are also other techniques in modern compilers and operating system kernels to mitigate vulnerabilities related to stack, like:
```